# Session 2 — Python Libraries: NumPy & Pandas

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/tech4alltraining/aiml/blob/main/mlai-genai-internship/student/notebooks/session-02-numpy-pandas.ipynb)

**ML/AI & GenAI Internship** · NumPy Fundamentals · Pandas · Data Loading & Exploration · EDA

---

## How to use this notebook

Run each cell with `Shift+Enter`. Read the explanation **before** the cell, run it, then look at the output.

| # | Topic |
|---|---|
| 1 | NumPy fundamentals |
| 2 | Pandas for data handling |
| 3 | Data loading and exploration |
| 4 | Exploratory Data Analysis |

Exercises, MCQs and tasks: [Session 2 guide](../sessions/session-02-numpy-pandas.md).

In [1]:
import numpy as np
import pandas as pd

# Every dataset loads straight from a URL - nothing to download.
BASE = "https://raw.githubusercontent.com/tech4alltraining/aiml/refs/heads/main/datasets/"

print("NumPy :", np.__version__)
print("Pandas:", pd.__version__)

NumPy : 2.4.0
Pandas: 2.3.3


---
# 1. NumPy fundamentals

🧠 **The shopping bag and the egg tray.** A list is a bag — to act on everything you pull items out one at a time. An array is an egg tray — every slot is the same kind of thing, so the computer acts on **all of them at once**.

In [2]:
prices_list  = [100, 200, 300]
prices_array = np.array([100, 200, 300])

print("list  * 2 =", prices_list * 2)
print("            ^ the BAG was duplicated")
print()
print("array * 2 =", prices_array * 2)
print("            ^ every ELEMENT was doubled  <- what we want")

list  * 2 = [100, 200, 300, 100, 200, 300]
            ^ the BAG was duplicated

array * 2 = [200 400 600]
            ^ every ELEMENT was doubled  <- what we want


In [3]:
a = np.array([1, 2, 3, 4, 5])
b = np.array([[1, 2, 3],
              [4, 5, 6]])

print("a.shape :", a.shape, "  5 elements, 1 dimension")
print("b.shape :", b.shape, " 2 rows, 3 columns")
print("a.dtype :", a.dtype)

print("\na + 10 :", a + 10)      # no loop anywhere
print("a ** 2 :", a ** 2)
print("\nb transposed:\n", b.T)
print("\nreshape:\n", np.arange(12).reshape(3, 4))

a.shape : (5,)   5 elements, 1 dimension
b.shape : (2, 3)  2 rows, 3 columns
a.dtype : int64

a + 10 : [11 12 13 14 15]
a ** 2 : [ 1  4  9 16 25]

b transposed:
 [[1 4]
 [2 5]
 [3 6]]

reshape:
 [[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]]


In [4]:
data = np.array([23, 45, 12, 67, 34, 89, 21])

print("mean  :", data.mean().round(2))
print("median:", np.median(data))
print("std   :", data.std().round(2))
print("min/max:", data.min(), data.max())

# BOOLEAN MASKING - the idea behind every Pandas filter you will write
mask = data > 30
print("\ndata > 30  :", mask)
print("data[mask] :", data[mask], " only the True positions")
print("how many   :", mask.sum(), " (True counts as 1)")

mean  : 41.57
median: 34.0
std   : 25.66
min/max: 12 89

data > 30  : [False  True False  True  True  True False]
data[mask] : [45 67 34 89]  only the True positions
how many   : 4  (True counts as 1)


In [5]:
# Reproducible randomness: fix the seed so results can be compared.
rng = np.random.default_rng(seed=42)
print("seed 42:", rng.integers(1, 100, size=5))

rng2 = np.random.default_rng(seed=42)
print("again  :", rng2.integers(1, 100, size=5), " <- identical")

seed 42: [ 9 77 65 44 43]
again  : [ 9 77 65 44 43]  <- identical


---
# 2. Pandas for data handling

🧠 **A spreadsheet that takes orders.** Everything you can do in Excel has a one-line Pandas equivalent.

In [6]:
df = pd.read_csv(BASE + "regression/cardekho_dataset.csv")
print("Loaded", df.shape[0], "cars,", df.shape[1], "columns")
df.head(3)

Loaded 15411 cars, 14 columns


,Unnamed: 0,car_name,brand,model,vehicle_age,km_driven,seller_type,fuel_type,transmission_type,mileage,engine,max_power,seats,selling_price
0,0,Maruti Alto,Maruti,Alto,9,120000,Individual,Petrol,Manual,19.7,796,46.3,5,120000
1,1,Hyundai Grand,Hyundai,Grand,5,20000,Individual,Petrol,Manual,18.9,1197,82.0,5,550000
2,2,Hyundai i20,Hyundai,i20,11,60000,Individual,Petrol,Manual,17.0,1197,80.0,5,215000


In [7]:
# ONE name in [] gives a Series. A LIST of names gives a DataFrame.
print("one column  ->", type(df["selling_price"]).__name__)
print("two columns ->", type(df[["brand", "selling_price"]]).__name__)

df[["car_name", "brand", "selling_price", "fuel_type"]].head()

one column  -> Series
two columns -> DataFrame


,car_name,brand,selling_price,fuel_type
0,Maruti Alto,Maruti,120000,Petrol
1,Hyundai Grand,Hyundai,550000,Petrol
2,Hyundai i20,Hyundai,215000,Petrol
3,Maruti Alto,Maruti,226000,Petrol
4,Ford Ecosport,Ford,570000,Diesel


In [8]:
# FILTERING: use & for AND, | for OR, and wrap EACH condition in brackets.
# Plain `and` / `or` do not work on whole columns.

expensive = df[df["selling_price"] > 1000000]
recent_petrol = df[(df["vehicle_age"] < 5) & (df["fuel_type"] == "Petrol")]

print("cars above 10 lakh :", len(expensive))
print("recent petrol cars :", len(recent_petrol))
recent_petrol[["car_name", "vehicle_age", "selling_price"]].head()

cars above 10 lakh : 2510
recent petrol cars : 3098


,car_name,vehicle_age,selling_price
7,Maruti Wagon R,3,410000
8,Hyundai Venue,2,1050000
9,Maruti Swift,4,511000
12,Mini Cooper,4,3250000
28,Hyundai i20,3,700000


In [9]:
# GROUPING is the single most useful Pandas skill.
print("Average price by fuel type:")
print(df.groupby("fuel_type")["selling_price"].mean().round(0).sort_values(ascending=False))

print("\nSeveral questions at once with agg():")
print(df.groupby("brand")["selling_price"].agg(["count", "mean", "max"]).head(5).round(0))

Average price by fuel type:
fuel_type
Electric    1853500.0
Diesel      1000469.0
Petrol       572862.0
CNG          417688.0
LPG          206273.0
Name: selling_price, dtype: float64

Several questions at once with agg():
         count        mean       max
brand                               
Audi       192   1966865.0   6800000
BMW        439   2693827.0   8500000
Bentley      3   9266667.0  14500000
Datsun     170    320518.0    650000
Ferrari      1  39500000.0  39500000


In [10]:
# A new column from existing ones
df["price_lakhs"] = (df["selling_price"] / 100000).round(2)

print("Three most expensive:")
print(df.sort_values("selling_price", ascending=False)[["car_name", "price_lakhs"]].head(3))

print("\nHow many of each fuel type:")
print(df["fuel_type"].value_counts())
print("\nAs proportions:")
print(df["fuel_type"].value_counts(normalize=True).round(3))

Three most expensive:
                  car_name  price_lakhs
3799     Ferrari GTC4Lusso        395.0
10969    Rolls-Royce Ghost        242.0
1172   Bentley Continental        145.0

How many of each fuel type:
fuel_type
Petrol      7643
Diesel      7419
CNG          301
LPG           44
Electric       4
Name: count, dtype: int64

As proportions:
fuel_type
Petrol      0.496
Diesel      0.481
CNG         0.020
LPG         0.003
Electric    0.000
Name: proportion, dtype: float64


---
# 3. Data loading and exploration

**Five commands, on every single new dataset, in this order.**

In [11]:
loans = pd.read_csv(BASE + "loan_data_10k.csv")

print("=== 1. How big is it? ===")
print(loans.shape)

=== 1. How big is it? ===
(10000, 14)


In [12]:
# 2. Do the values look sensible?
loans.head()

,person_age,person_gender,person_education,person_income,person_emp_exp,person_home_ownership,loan_amnt,loan_intent,loan_int_rate,loan_percent_income,cb_person_cred_hist_length,credit_score,previous_loan_defaults_on_file,loan_status
0,29,female,Associate,39704.0,8,RENT,10800.0,MEDICAL,13.57,0.27,5,607,No,1
1,28,male,Associate,36889.0,5,RENT,7796.0,PERSONAL,14.04,0.21,10,668,Yes,0
2,26,male,Associate,97163.0,2,RENT,20000.0,EDUCATION,10.05,0.21,6,643,No,0
3,41,male,Doctorate,65605.0,23,RENT,10000.0,MEDICAL,13.23,0.15,17,600,Yes,0
4,24,male,Bachelor,37030.0,1,RENT,4600.0,MEDICAL,11.01,0.12,2,657,Yes,0


In [13]:
# 3. Column names, types, non-null counts.
# Watch for numbers stored as 'object' - that means text, and it will
# break your model later.
loans.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   person_age                      10000 non-null  int64  
 1   person_gender                   9999 non-null   object 
 2   person_education                10000 non-null  object 
 3   person_income                   10000 non-null  float64
 4   person_emp_exp                  10000 non-null  int64  
 5   person_home_ownership           10000 non-null  object 
 6   loan_amnt                       10000 non-null  float64
 7   loan_intent                     10000 non-null  object 
 8   loan_int_rate                   9999 non-null   float64
 9   loan_percent_income             9999 non-null   float64
 10  cb_person_cred_hist_length      10000 non-null  int64  
 11  credit_score                    10000 non-null  int64  
 12  previous_loan_defaults_on_file  1

In [14]:
# 4. Ranges and outliers. Compare 75% against max.
loans.describe().T[["mean", "min", "50%", "75%", "max"]].round(2)

,mean,min,50%,75%,max
person_age,27.70,20.00,26.00,30.00,144.00
person_income,72289.64,8000.00,60954.00,87421.50,2448661.00
person_emp_exp,5.34,0.00,4.00,8.00,121.00
loan_amnt,10283.86,1000.00,8500.00,14000.00,2500000.00
loan_int_rate,11.71,5.42,11.39,13.99,101.00
loan_percent_income,0.16,0.00,0.14,0.23,0.62
cb_person_cred_hist_length,5.85,2.00,4.00,8.00,30.00
credit_score,632.36,418.00,639.00,670.00,807.00
loan_status,0.50,0.00,0.50,1.00,1.00


In [15]:
# 5. What is missing, and where?
missing = loans.isnull().sum()
print(missing[missing > 0] if missing.sum() else "No missing values.")

person_gender          1
loan_int_rate          1
loan_percent_income    1
dtype: int64


### Twelve rows you can check by eye

Before working with 10,000 rows, work with one you can hold in your head.

In [16]:
small = pd.read_csv(BASE + "prepreprocessing/pre_data.csv")
small

,Country,Age,Salary,Purchased
0,France,44.0,72000.0,No
1,Spain,27.0,48000.0,Yes
2,NaN,30.0,54000.0,No
3,Spain,38.0,61000.0,No
4,Germany,40.0,NaN,Yes
5,France,35.0,58000.0,Yes
6,Spain,NaN,52000.0,No
7,France,48.0,79000.0,Yes
8,Germany,50.0,83000.0,No
9,France,37.0,67000.0,Yes


In [17]:
print("--- missing values per column ---")
print(small.isnull().sum())

print("\n--- duplicate rows ---")
print(small[small.duplicated(keep=False)])

print("\n--- numeric summary: compare 75% against max ---")
print(small.describe().round(1))

--- missing values per column ---
Country      1
Age          1
Salary       1
Purchased    0
dtype: int64

--- duplicate rows ---
    Country   Age   Salary Purchased
8   Germany  50.0  83000.0        No
10  Germany  50.0  83000.0        No

--- numeric summary: compare 75% against max ---
        Age     Salary
count  11.0       11.0
mean   40.8   196090.9
std     8.2   432635.3
min    27.0    48000.0
25%    36.0    56000.0
50%    40.0    67000.0
75%    49.0    81000.0
max    50.0  1500000.0


**✅ Two things to notice in that last table.**

1. Salary `75%` is **81,000** but `max` is **1,500,000**. *A max many times the 75th percentile is the signature of an outlier.*
2. Salary `mean` is about **196,091** but the median (`50%`) is **67,000**. One row moved the average by nearly 130,000.

**That second point is the concrete reason we fill missing numbers with the median, not the mean.**

---
# 4. Exploratory Data Analysis

🧠 **The doctor's check-up.** A doctor does not prescribe the moment you walk in. Skipping EDA and jumping to `model.fit()` is prescribing without examining the patient.

In [18]:
# Is the target balanced? This decides which metrics you can trust.
print(loans["loan_status"].value_counts())
print()
print(loans["loan_status"].value_counts(normalize=True).round(3))

loan_status
1    5000
0    5000
Name: count, dtype: int64

loan_status
1    0.5
0    0.5
Name: proportion, dtype: float64


In [19]:
# Which numeric columns relate to the target?
#
# Near +1 or -1 -> moves WITH the target
# Near 0        -> no LINEAR relationship ON ITS OWN
#
# NOTE: correlation is NOT causation. Ice cream sales correlate with
# drowning deaths - hot weather causes both.

loans.corr(numeric_only=True)["loan_status"].sort_values().round(3)

person_income                -0.210
person_age                   -0.039
person_emp_exp               -0.039
cb_person_cred_hist_length   -0.033
credit_score                 -0.002
loan_amnt                     0.042
loan_int_rate                 0.363
loan_percent_income           0.405
loan_status                   1.000
Name: loan_status, dtype: float64

In [20]:
# A group comparison often says more than a correlation number.
print("Average credit score by outcome:")
print(loans.groupby("loan_status")["credit_score"].mean().round(1))

print("\nAverage loan-to-income by outcome:")
loans["loan_to_income"] = loans["loan_amnt"] / loans["person_income"]
print(loans.groupby("loan_status")["loan_to_income"].mean().round(4))

Average credit score by outcome:
loan_status
0    632.5
1    632.2
Name: credit_score, dtype: float64

Average loan-to-income by outcome:
loan_status
0    0.1226
1    0.2124
Name: loan_to_income, dtype: float64


### 🔁 Change one thing

Swap `loan_data_10k.csv` for `classification/diabetes_prediction_dataset.csv` and run the five commands again.

**Then answer:** is that target balanced? What does your answer mean for whether you can trust accuracy as a metric?

In [21]:
# Your turn.
dia = pd.read_csv(BASE + "classification/diabetes_prediction_dataset.csv")
print(dia.shape)
print(dia["diabetes"].value_counts(normalize=True).round(3))

(100000, 9)
diabetes
0    0.915
1    0.085
Name: proportion, dtype: float64


---
# ✅ Before you move on

- [ ] Why an array beats a list — the egg-tray analogy
- [ ] Filtering an array with a boolean mask, no loop
- [ ] Loading a CSV from a URL and from a file
- [ ] The five commands, on every new dataset
- [ ] Filtering a DataFrame on two conditions with `&`
- [ ] `groupby`, and explaining the result in a sentence
- [ ] Why `df.dropna()` alone changes nothing
- [ ] Spotting an outlier from `describe()` by comparing `75%` and `max`
- [ ] Why the median beats the mean for filling gaps
- [ ] That correlation is not causation

**Exercises, MCQs and tasks:** [Session 2 guide](../sessions/session-02-numpy-pandas.md)

| | |
|---|---|
| **Previous** | [Session 1 — Python Refresher](../sessions/session-01-python-refresher.md) |
| **Next** | [Sessions index](../sessions/README.md) |
| **Stuck?** | [Troubleshooting](../troubleshooting.md) |